File: Diagrammatize.Design.Pattern/code/cxx/ChainOfResponsibility/A4/Support.cpp
#include "Support.hpp"
#include <iostream>

Support::Support(const std::string& name_): name(name_), next(nullptr) {
}

Support* Support::setNext(Support* next_) {
    next = next_;
    return next;
}

void Support::support(Trouble* trouble) {
    for (Support* obj = this; true; obj = obj->next) {
        if (obj->resolve(trouble)) {
            obj->done(trouble);
            break;
        } else if (obj->next == nullptr) {
            obj->fail(trouble);
            break;
        }
    }
}

std::string Support::toString() const {
    return "[" + name + "]";
}

void Support::done(Trouble* trouble) {
    std::cout << *trouble << " is resolved by " << *this << "." << std::endl;
}

void Support::fail(Trouble* trouble) {
    std::cout << *trouble << " cannot be resolved." << std::endl;
}

std::ostream& operator<<(std::ostream& os, const Support& s) {
    os << s.toString();
    return os;
}

File: Diagrammatize.Design.Pattern/code/python/ChainOfResponsibility/A4/Support.py
import abc

class Support(abc.ABC):
    def __init__(self, name):
        self.name = name
        self.next = None

    def setNext(self, next_):
        self.next = next_
        return next_

    def support(self, trouble):
        obj = self
        while True:
            if obj.resolve(trouble):
                obj.done(trouble)
                break
            elif obj.next is None:
                obj.fail(trouble)
                break
            obj = obj.next

    def __str__(self):
        return "[{}]".format(self.name)

    @abc.abstractmethod
    def resolve(self, trouble):
        pass

    def done(self, trouble):
        print("{} is resolved by {}.".format(trouble, self))

    def fail(self, trouble):
        print("{} cannot be resolved.".format(trouble))
